File: app/src/main/java/com/example/projetoestagioaiko/activity/LinhasActivity.java
package com.example.projetoestagioaiko.activity;

import android.annotation.SuppressLint;
import android.os.Bundle;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.projetoestagioaiko.api.OlhoVivoApiService;
import com.example.projetoestagioaiko.R;
import com.example.projetoestagioaiko.api.RetrofitClient;
import com.example.projetoestagioaiko.adapters.LinhaOnibusAdapter;
import com.example.projetoestagioaiko.models.LinhaOnibus;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class LinhasActivity extends AppCompatActivity {
    private RecyclerView recyclerView;
    private LinhaOnibusAdapter adapter;
    private OlhoVivoApiService olhoVivoApiService;
    private List<LinhaOnibus> listaLinhas;

    @SuppressLint("MissingInflatedId")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_linhas);

        recyclerView = findViewById(R.id.recycler_view_linhas);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        olhoVivoApiService = RetrofitClient.getClient("https://api.olhovivo.sptrans.com.br/").create(OlhoVivoApiService.class);
        listaLinhas = new ArrayList<>();
        adapter = new LinhaOnibusAdapter(listaLinhas);
        recyclerView.setAdapter(adapter);

        carregarLinhas();
    }

    private void carregarLinhas(){
        olhoVivoApiService.obterLinhas().enqueue(new Callback<List<LinhaOnibus>>() {
            @SuppressLint("NotifyDataSetChanged")
            @Override
            public void onResponse(Call<List<LinhaOnibus>> call, Response<List<LinhaOnibus>> response) {
                if(response.isSuccessful() && response.body() != null){
                    listaLinhas.clear();
                    listaLinhas.addAll(response.body());
                    adapter.notifyDataSetChanged();
                }
            }

            @Override
            public void onFailure(Call<List<LinhaOnibus>> call, Throwable t) {

            }
        });
    }
}

File: app/src/main/java/br/dev/saed/saedrastreamentosapi/activities/LinhasActivity.kt
package br.dev.saed.saedrastreamentosapi.activities

import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import br.dev.saed.saedrastreamentosapi.R
import br.dev.saed.saedrastreamentosapi.adapters.recyclerview.LinhaAdapter
import br.dev.saed.saedrastreamentosapi.apis.OlhoVivoAPI
import br.dev.saed.saedrastreamentosapi.apis.RetrofitHelper
import br.dev.saed.saedrastreamentosapi.databinding.ActivityLinhasBinding
import br.dev.saed.saedrastreamentosapi.models.Linha
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import retrofit2.Retrofit

class LinhasActivity : AppCompatActivity() {

    private lateinit var binding: ActivityLinhasBinding
    private lateinit var retrofit: Retrofit
    private lateinit var linhaAdapter: LinhaAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityLinhasBinding.inflate(layoutInflater)
        setContentView(binding.root)
        retrofit = RetrofitHelper.retrofit
        inicializarComponentes()
        carregarRV()
    }

    private fun inicializarComponentes() {
        binding.buttonProcurarLinhas.setOnClickListener {
            val texto = binding.editTextLinha.text.toString()
            if (texto.isEmpty()) {
                binding.editTextLinha.error = "Obrigatório!"
            } else {
                CoroutineScope(Dispatchers.IO).launch {
                    popularRV(texto, linhaAdapter)
                }
            }
        }

        binding.fabInformacoesLinhas.setOnClickListener {
            val alertDialog = AlertDialog.Builder(this)
            alertDialog
                .setTitle("Linhas")
                .setMessage(R.string.text_info_linhas)
                .setPositiveButton("OK") { _, _ -> }
                .create()
                .show()
        }
    }

    private fun carregarRV() {
        linhaAdapter = LinhaAdapter()
        binding.rvLinhas.adapter = linhaAdapter
        binding.rvLinhas.layoutManager =
            LinearLayoutManager(applicationContext, RecyclerView.VERTICAL, false)
    }

    private fun popularRV(texto: String, linhaAdapter: LinhaAdapter) {
        CoroutineScope(Dispatchers.IO).launch {
            val listaLinhas = async {
                buscarLinhas(texto)
            }.await()
            withContext(Dispatchers.Main) {
                linhaAdapter.atualizar(listaLinhas)
            }
        }
    }

    private suspend fun buscarLinhas(texto: String): List<Linha> {
        val linhas = mutableListOf<Linha>()

        try {
            val resultado = retrofit.create(OlhoVivoAPI::class.java).buscarLinhas(texto)
            if (resultado.isSuccessful) {
                val linha = resultado.body()
                linha?.forEach {
                    linhas.add(it)
                }
            }
        } catch (e: Exception) {
            withContext(Dispatchers.Main) {
                Toast.makeText(
                    applicationContext,
                    R.string.text_no_internet,
                    Toast.LENGTH_LONG
                ).show()
            }
        }
        return linhas
    }
}